[automation] Docker artifact building errors, but it doesn't cause a build failure

We've seen an error when building docker images where a failure is not really stopping the build process. Instead, it continues to build images, which results in images being built with incomplete data or missing pieces. For example, the error can be seen in https://logstash-ci.elastic.co/job/elastic+logstash+7.17+multijob-dra-docker-x86_64/6/console, with the following example.

 ```
[docker_oss] Building OSS docker image
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
LOCAL_PY_VER=`python3 --version 2>&1`&&\
echo "Was using $LOCAL_PY_VER" &&\
eval "$(pyenv init -)" && eval "$(pyenv init --path)" &&\
pyenv install --skip-existing 3.6.13 &&\
pyenv local 3.6.13 &&\
python3 -mvenv venv && \
for i in 0 1 2 3 4 5; do sleep ""; pip install -r requirements.txt && break; done &&\
touch venv
Was using Python 3.10.6
Downloading Python-3.6.13.tar.xz...
-> https://www.python.org/ftp/python/3.6.13/Python-3.6.13.tar.xz
Installing Python-3.6.13...

BUILD FAILED (Ubuntu 22.04 using python-build 20180424)

Inspect or clean up the working tree at /tmp/python-build.20221026005428.17793
Results logged to /tmp/python-build.20221026005428.17793.log

Last 10 log lines:
	case upgrade in \
		upgrade) ensurepip="--upgrade" ;; \
		install|*) ensurepip="" ;; \
	esac; \
	 ./python -E -m ensurepip \
		$ensurepip --root=/ ; \
fi
Segmentation fault
make[1]: *** [Makefile:1102: install] Error 139
make[1]: Leaving directory '/tmp/python-build.20221026005428.17793/Python-3.6.13'
make: *** [Makefile:187: venv] Error 1
```

However, the build continues and even errors again":

```
Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8 -Dsun.jnu.encoding=UTF8
unsupported Java version "11", defaulting to 1.7
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.jruby.util.io.FilenoUtil to method sun.nio.ch.SelChImpl.getFD()
WARNING: Please consider reporting this to the maintainers of org.jruby.util.io.FilenoUtil
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
Using JAVA_HOME defined java: /var/lib/jenkins/.java/openjdk-11.0.2-linux
DEPRECATION: The use of JAVA_HOME is now deprecated and will be removed starting from 8.0. Please configure LS_JAVA_HOME instead.
warning: ignoring JAVA_TOOL_OPTIONS=-Dfile.encoding=UTF8 -Dsun.jnu.encoding=UTF8
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
Skipping bundler install...
Building logstash-core using gradle
./gradlew assemble
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
WARNING: Unknown module: org.jruby.dist specified to --add-opens
WARNING: Unknown module: org.jruby.dist specified to --add-opens
WARNING: Unknown module: org.jruby.dist specified to --add-opens
WARNING: Unknown module: org.jruby.dist specified to --add-opens
To honour the JVM settings for this build a single-use Daemon process will be forked. See https://docs.gradle.org/7.2/userguide/gradle_daemon.html#sec:disabling_the_daemon.
Daemon will be stopped at the end of the build 

[...]

LOCAL_PY_VER=`python3 --version 2>&1`&&\
echo "Was using $LOCAL_PY_VER" &&\
eval "$(pyenv init -)" && eval "$(pyenv init --path)" &&\
pyenv install --skip-existing 3.6.13 &&\
pyenv local 3.6.13 &&\
python3 -mvenv venv && \
for i in 0 1 2 3 4 5; do sleep ""; pip install -r requirements.txt && break; done &&\
touch venv
Was using Python 3.10.6
Downloading Python-3.6.13.tar.xz...
-> https://www.python.org/ftp/python/3.6.13/Python-3.6.13.tar.xz
Installing Python-3.6.13...

BUILD FAILED (Ubuntu 22.04 using python-build 20180424)

Inspect or clean up the working tree at /tmp/python-build.20221026005720.29757
Results logged to /tmp/python-build.20221026005720.29757.log

Last 10 log lines:
	case upgrade in \
		upgrade) ensurepip="--upgrade" ;; \
		install|*) ensurepip="" ;; \
	esac; \
	 ./python -E -m ensurepip \
		$ensurepip --root=/ ; \
fi
Segmentation fault
make[1]: *** [Makefile:1102: install] Error 139
make[1]: Leaving directory '/tmp/python-build.20221026005720.29757/Python-3.6.13'
make: *** [Makefile:187: venv] Error 1
```

Even though the error should have caused the build to fail, it succeeded: https://logstash-ci.elastic.co/job/elastic+logstash+7.17+multijob-dra-docker-x86_64/6/ 

This ultimately resulted in missing fails, erroring with: https://logstash-ci.elastic.co/job/elastic+logstash+7.17+multijob-dra-upload/11/console

See for full story: https://logstash-ci.elastic.co/view/DRA/job/elastic+logstash+7.17+dra-snapshot/33/

--- a/rakelib/artifacts.rake
+++ b/rakelib/artifacts.rake
@@ -126,6 +126,14 @@ namespace "artifact" do
     result
   end
 
+  # execute Kernel#system call,checking the exist status of the executed command and eventually reporting as exception
+  def safe_system(*args)
+    if !system(*args)
+      status = $?
+      raise "Got exit status #{status.exitstatus} attempting to execute #{args.inspect}!"
+    end
+  end
+
   desc "Generate rpm, deb, tar and zip artifacts"
   task "all" => ["prepare", "build"]
   task "docker_only" => ["prepare", "build_docker_full", "build_docker_oss", "build_docker_ubi8"]
@@ -139,7 +147,7 @@ namespace "artifact" do
     create_archive_pack(license_details, "arm64", "linux", "darwin")
 
     #without JDK
-    raise "Gradle failed to execute bootstrap" unless system("./gradlew bootstrap") #force the build of Logstash jars
+    safe_system("./gradlew bootstrap") #force the build of Logstash jars
     @bundles_jdk = false
     build_tar(*license_details, platform: '-no-jdk')
     build_zip(*license_details, platform: '-no-jdk')
@@ -153,7 +161,7 @@ namespace "artifact" do
   end
 
   def create_single_archive_pack(os_name, arch, license_details)
-    raise "Gradle failed to execute copyJdk" unless system("./gradlew copyJdk -Pjdk_bundle_os=#{os_name} -Pjdk_arch=#{arch}")
+    safe_system("./gradlew copyJdk -Pjdk_bundle_os=#{os_name} -Pjdk_arch=#{arch}")
     if arch == 'arm64'
       arch = 'aarch64'
     end
@@ -165,19 +173,15 @@ namespace "artifact" do
     when "darwin"
       build_tar(*license_details, platform: "-darwin-#{arch}")
     end
-    raise "Gradle failed to execute deleteLocalJdk" unless system("./gradlew deleteLocalJdk -Pjdk_bundle_os=#{os_name}")
+    safe_system("./gradlew deleteLocalJdk -Pjdk_bundle_os=#{os_name}")
   end
 
   # Create an archive pack using settings appropriate for the running machine
   def create_local_archive_pack(bundle_jdk)
     @bundles_jdk = bundle_jdk
-    if bundle_jdk
-      raise "Gradle failed to execute copyJdk" unless system("./gradlew copyJdk")
-    end
+    safe_system("./gradlew copyJdk") if bundle_jdk
     build_tar('ELASTIC-LICENSE')
-    if bundle_jdk
-      raise "Gradle failed to execute deleteLocalJdk" unless system("./gradlew deleteLocalJdk")
-    end
+    safe_system("./gradlew deleteLocalJdk") if bundle_jdk
   end
 
 
@@ -201,7 +205,7 @@ namespace "artifact" do
 
     #without JDK
     @bundles_jdk = false
-    raise "Gradle failed to execute bootstrap" unless system("./gradlew bootstrap") #force the build of Logstash jars
+    safe_system("./gradlew bootstrap") #force the build of Logstash jars
     build_tar(*license_details, platform: '-no-jdk')
     build_zip(*license_details, platform: '-no-jdk')
   end
@@ -218,7 +222,7 @@ namespace "artifact" do
 
     #without JDKs
     @bundles_jdk = false
-    raise "Gradle failed to execute deleteLocalJdk bootstrap" unless system("./gradlew bootstrap") #force the build of Logstash jars
+    safe_system("./gradlew bootstrap") #force the build of Logstash jars
     package("centos")
   end
 
@@ -234,7 +238,7 @@ namespace "artifact" do
 
     #without JDKs
     @bundles_jdk = false
-    raise "Gradle failed to execute bootstrap" unless system("./gradlew bootstrap") #force the build of Logstash jars
+    safe_system("./gradlew bootstrap") #force the build of Logstash jars
     package("centos", :oss)
   end
 
@@ -251,7 +255,7 @@ namespace "artifact" do
 
     #without JDKs
     @bundles_jdk = false
-    raise "Gradle failed to execute bootstrap" unless system("./gradlew bootstrap") #force the build of Logstash jars
+    safe_system("./gradlew bootstrap") #force the build of Logstash jars
     package("ubuntu")
   end
 
@@ -267,7 +271,7 @@ namespace "artifact" do
 
     #without JDKs
     @bundles_jdk = false
-    raise "Gradle failed to execute bootstrap" unless system("./gradlew bootstrap") #force the build of Logstash jars
+    safe_system("./gradlew bootstrap") #force the build of Logstash jars
     package("ubuntu", :oss)
   end
 
@@ -532,9 +536,9 @@ namespace "artifact" do
   end
 
   def package_with_jdk(platform, jdk_arch, variant=:standard)
-    raise "Gradle failed to execute copyJdk" unless system("./gradlew copyJdk -Pjdk_bundle_os=linux -Pjdk_arch=#{jdk_arch}")
+    safe_system("./gradlew copyJdk -Pjdk_bundle_os=linux -Pjdk_arch=#{jdk_arch}")
     package(platform, variant, true, jdk_arch)
-    raise "Gradle failed to execute deleteLocalJdk" unless system('./gradlew deleteLocalJdk -Pjdk_bundle_os=linux')
+    safe_system('./gradlew deleteLocalJdk -Pjdk_bundle_os=linux')
   end
 
   def package(platform, variant=:standard, bundle_jdk=false, jdk_arch='x86_64')
@@ -750,7 +754,7 @@ namespace "artifact" do
       "BUILD_DATE" => BUILD_DATE
     }
     Dir.chdir("docker") do |dir|
-      raise "build-from-local-#{flavor}-artifacts Make invocation errored" unless system(env, "make build-from-local-#{flavor}-artifacts")
+      safe_system(env, "make build-from-local-#{flavor}-artifacts")
     end
   end
 
@@ -762,7 +766,7 @@ namespace "artifact" do
       "BUILD_DATE" => BUILD_DATE
     }
     Dir.chdir("docker") do |dir|
-      raise "public-dockerfiles_#{flavor} Make invocation errored" unless system(env, "make public-dockerfiles_#{flavor}")
+      safe_system(env, "make public-dockerfiles_#{flavor}")
       puts "Dockerfiles created in #{env['ARTIFACTS_DIR']}"
     end
   end

